<a href="https://colab.research.google.com/github/kaytheteenwriter-prog/Leano-Khanyisa-Giyose-ML-Model-to-Web-App-Lab/blob/main/Copy_of_ML_Model_to_Web_App_Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practical Lab: From Machine Learning Model to Web App Development

## Dataset Overview

The selected **Housing.csv** dataset is ideal for teaching deployment. It contains **545 records** and **13 columns**, mixing numerical and categorical features for real-world preprocessing.

### Dataset Features

| Feature | Type | Meaning |
| --- | --- | --- |
| **price** | numeric | Target variable (house price) |
| **area** | numeric | House size |
| **bedrooms** | numeric | Number of bedrooms |
| **bathrooms** | numeric | Number of bathrooms |
| **stories** | numeric | Number of floors |
| **mainroad** | categorical | Access to main road |
| **guestroom** | categorical | Guest room availability |
| **basement** | categorical | Basement availability |
| **hotwaterheating** | categorical | Hot water heating |
| **airconditioning** | categorical | Air conditioning |
| **parking** | numeric | Number of parking spaces |
| **prefarea** | categorical | Preferred area |
| **furnishingstatus** | categorical | Furnishing type |

---

## 1. Project Title

**House Price Prediction Web Application**
This project predicts the price of a house based on its features. Users will enter house characteristics and the system will return a predicted price.

## 2. Machine Learning Pipeline

Explain the workflow clearly to students:
**Dataset $\rightarrow$ Data Cleaning $\rightarrow$ Model Training $\rightarrow$ Save Model $\rightarrow$ Build Web App $\rightarrow$ Deploy**

---

## 3. Step 1: Load Dataset in Google Colab

In [3]:
import pandas as pd

data = pd.read_csv("/content/Housing.csv")
data.head()

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


*Explain that each row represents one house listing.*

## 4. Step 2: Inspect the Data

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


Students will see integers, categorical variables (yes/no), and furnishing categories.

## 5. Step 3: Convert Categorical Variables

Machine learning models cannot understand text like "yes" or "no". We convert them to numbers.

In [5]:
binary_cols = [
    'mainroad','guestroom','basement',
    'hotwaterheating','airconditioning','prefarea'
]

for col in binary_cols:
    data[col] = data[col].map({'yes':1,'no':0})

## 6. Step 4: Handle Furnishing Status

This column has 3 categories: **furnished**, **semi-furnished**, and **unfurnished**.

In [6]:
data['furnishingstatus'] = data['furnishingstatus'].map({
    'furnished':2,
    'semi-furnished':1,
    'unfurnished':0
})

*Explain that this process is called encoding categorical variables.*

## 7. Step 5: Separate Features and Target

In [7]:
X = data.drop('price', axis=1)
y = data['price']

* **Features:** Describe the house.
* **Target:** Price is what we want to predict.

## 8. Step 6: Split the Data

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

*Explain: 80% training, 20% testing.*

## 9. Step 7: Train the Model

In [9]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

*Explain that the model learns relationships between house features and price.*

## 10. Step 8: Evaluate Model

In [10]:
from sklearn.metrics import mean_absolute_error

predictions = model.predict(X_test)
mae = mean_absolute_error(y_test, predictions)

print("Mean Absolute Error:", mae)

Mean Absolute Error: 979679.6912959901


## *Interpreting the Mean Absolute Error (MAE)*

- The Mean Absolute Error (MAE) measures the average difference between the actual house prices and the prices predicted by the model. It is calculated by taking the absolute value of the difference between the predicted value and the true value for each observation, and then averaging those differences across all predictions.

- In this project, the MAE value obtained was 979,679.69. This means that, on average, the model’s predicted house prices differ from the true prices by approximately 979,679 units of currency. In simpler terms, the model typically makes an error of about one million when estimating the price of a house.

- To understand this more clearly, consider an example. If the model predicts a house price of 4,500,000, the actual price may realistically fall somewhere within about ±979,000 of that value. For instance, the true price might be around 3,520,000 or 5,480,000. The MAE represents the typical size of these prediction errors across the entire dataset.

- It is important to note that MAE is expressed in the same units as the target variable, which in this dataset is the house price. Since house prices in the dataset range from roughly 3,000,000 to over 13,000,000, an average prediction error of about 1,000,000 represents roughly 8–15% of the typical house price range.

- For a simple baseline machine learning model, this level of error is reasonable. More advanced models, additional features, or further data preprocessing could potentially reduce the prediction error and improve the model’s accuracy.

- Overall, the MAE helps us understand how accurate our model is by providing a clear and interpretable measure of the typical prediction error.

*Students learn that evaluation measures model accuracy.*

## 11. Step 9: Save the Model

In [11]:
import joblib

joblib.dump(model, "house_model.pkl")

['house_model.pkl']

*This creates a portable model file used by the web application.*

## 12. Step 10: Download Model from Colab

In [12]:
from google.colab import files
files.download("house_model.pkl")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

---

## 13. Building the Streamlit App

Create a file named: `app.py`

## 14. Web App Code

In [13]:
!pip install streamlit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 60.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 123.8 MB/s eta 0:00:00


In [14]:
#We have not yet created the py file rather we just designed the flow of its contents, there is an additional step proceeding
import streamlit as st
import joblib
import numpy as np

model = joblib.load("house_model.pkl")

st.title("House Price Prediction")

area = st.number_input("Area")
bedrooms = st.number_input("Bedrooms")
bathrooms = st.number_input("Bathrooms")
stories = st.number_input("Stories")
parking = st.number_input("Parking")

mainroad = st.selectbox("Main Road", ["yes","no"])
guestroom = st.selectbox("Guest Room", ["yes","no"])
basement = st.selectbox("Basement", ["yes","no"])
airconditioning = st.selectbox("Air Conditioning", ["yes","no"])

if st.button("Predict Price"):

    features = np.array([[area,bedrooms,bathrooms,stories,
                          1 if mainroad=="yes" else 0,
                          1 if guestroom=="yes" else 0,
                          1 if basement=="yes" else 0,
                          0,
                          1 if airconditioning=="yes" else 0,
                          parking,
                          0,
                          1]])

    prediction = model.predict(features)

    st.success(f"Estimated Price: {prediction[0]}")

2026-03-07 20:39:17.437 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-03-07 20:39:17.611 
  command:

    streamlit run /usr/local/lib/python3.12/dist-packages/colab_kernel_launcher.py [ARGUMENTS]
2026-03-07 20:39:17.612 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-03-07 20:39:17.613 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-03-07 20:39:17.614 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-03-07 20:39:17.616 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-03-07 20:39:17.617 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2026-03-07 20:39:17.617 Thread 'MainThread': mi

In [15]:
#Method 2 (Create app.py from inside Jupyter / Colab)
#This step is to create the actual app file within our jupyter notebook i=environment
code = """
import streamlit as st
import joblib
import numpy as np

model = joblib.load("house_model.pkl")  #The machine learning model that has been trained

st.title("House Price Prediction")

area = st.number_input("Area")
bedrooms = st.number_input("Bedrooms")
bathrooms = st.number_input("Bathrooms")
stories = st.number_input("Stories")
parking = st.number_input("Parking")

mainroad = st.selectbox("Main Road", ["yes","no"])
guestroom = st.selectbox("Guest Room", ["yes","no"])
basement = st.selectbox("Basement", ["yes","no"])
airconditioning = st.selectbox("Air Conditioning", ["yes","no"])

if st.button("Predict Price"):

    features = np.array([[area,bedrooms,bathrooms,stories,
                          1 if mainroad=="yes" else 0,
                          1 if guestroom=="yes" else 0,
                          1 if basement=="yes" else 0,
                          0,
                          1 if airconditioning=="yes" else 0,
                          parking,
                          0,
                          1]])

    prediction = model.predict(features)

    st.success(f"Estimated Price: {prediction[0]}")
"""
with open("app.py", "w") as f:
    f.write(code)

#This will create:app.py in your notebook directory.
#Then Run the App
#Open terminal and run:
#streamlit run app.py
#Streamlit will launch the web app.


In [16]:
#Create requirements text file
#streamlit
#scikit-learn
#pandas
#numpy
#joblib

with open("requirements.txt", "w") as f:
  f.write("streamlit\nnumpy\nscikit-learn\njoblib")

## 15. Run the App and 16. Deployment

Upload the project to GitHub with:

1. `app.py`
2. `house_model.pkl`
3. `requirements.txt`

**requirements.txt content:**

```text
streamlit
scikit-learn
pandas
numpy
joblib

```

Deploy using: [https://streamlit.io/cloud](https://streamlit.io/cloud)

---

## What Students Achieve

By completing this project, students learn:

* Machine learning modeling
* Feature engineering
* Model serialization
* Web app creation
* Deployment
* Real-world ML workflow

This is exactly how AI systems move from **notebook $\rightarrow$ production**.

---



## 🚀 Final Deployment Process

Once you have satisfied the conditions mentioned above, follow these steps to take your model live:

* **Access Streamlit Cloud:** Navigate to [Streamlit Community Cloud](https://streamlit.io/cloud).
* **Create & Link:** Create an account and **link it to your GitHub profile**.
* **Prepare GitHub:** Create a new repository for the project and upload your necessary files (`app.py`, `house_model.pkl`, and `requirements.txt`).
* **Launch the App:** Link your new GitHub repository to Streamlit Cloud to build and deploy your application.

> **Deploy using the official portal:** [https://streamlit.io/cloud](https://streamlit.io/cloud)



# Deployment Instructions: Streamlit Community Cloud

Now you can go to Streamlit Community Cloud and deploy. Here’s the step-by-step, exactly what to click/do.

---

### **Before you go (quick checklist)**

Make sure these **3 files** are in the same folder (same repo root):

* `app.py`
* `house_model.pkl`
* `requirements.txt`

**If yes → proceed.**

---

### **Step-by-step: Deploy on Streamlit Cloud**

#### **1) Go to Streamlit Cloud**

Open your browser and go to:
[streamlit.io/cloud](https://streamlit.io/cloud)

(or search “Streamlit Community Cloud”)

#### **2) Sign in**

* Click **Sign in**
* Choose **Continue with GitHub**
* Authorize Streamlit (allow access)

#### **3) Create a GitHub repo (if you haven’t yet)**

In GitHub:

* Click **New repository**
* Name it something like: `house-price-app`
* Set it to **Public** (easiest for free deployment)
* Create repo

#### **4) Upload your files to the repo**

On the repo page:

* Click **Add file** → **Upload files**
* Upload: `app.py`, `house_model.pkl`, `requirements.txt`
* Click **Commit changes**

#### **5) Deploy from Streamlit Cloud**

Back on Streamlit Cloud:

* Click **New app** (or Create app)
* Choose **From existing repo**
* Select:
* **Repository:** `yourusername/house-price-app`
* **Branch:** `main` (default)
* **Main file path:** `app.py`



#### **6) Click Deploy**

* Click **Deploy**
* Wait for the build logs
* When done, Streamlit gives you a **public link** to your app

---

### **If deployment fails (most common reasons + fixes)**

**A) “ModuleNotFoundError…”**

* **Fix:** add the missing package to `requirements.txt` (often pandas). Recommended requirements:
* `streamlit`
* `numpy`
* `pandas`
* `scikit-learn`
* `joblib`


* Commit the updated file → Streamlit will rebuild automatically.

**B) “FileNotFoundError: house_model.pkl”**

* **Fix:** Ensure `house_model.pkl` is uploaded to the repo same level as `app.py`.
* And your code loads it like: `joblib.load("house_model.pkl")`

**C) App runs but predictions look weird**

* **Fix:** your app must include all dataset inputs (`hotwaterheating`, `prefarea`, `furnishingstatus`) or you must retrain the model using only the features you kept.

---



#Successfully created app

##access on this public link
[khanhousepriceapp-5suymypojzymmkvawv6aqw.streamlit.app](https://khanhousepriceapp-5suymypojzymmkvawv6aqw.streamlit.app)

## Github repository for project links:

[https://github.com/khannyasha/khan_house_price_app/tree/main](https://github.com/khannyasha/khan_house_price_app/tree/main)